Bricks/strings/piece.h

```diff
@@ -0,0 +1,201 @@
+/*******************************************************************************
+The MIT License (MIT)
+
+Copyright (c) 2015 Dmitry "Dima" Korolev <dmitry.korolev@gmail.com>
+
+Permission is hereby granted, free of charge, to any person obtaining a copy
+of this software and associated documentation files (the "Software"), to deal
+in the Software without restriction, including without limitation the rights
+to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+copies of the Software, and to permit persons to whom the Software is
+furnished to do so, subject to the following conditions:
+
+The above copyright notice and this permission notice shall be included in all
+copies or substantial portions of the Software.
+
+THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+SOFTWARE.
+*******************************************************************************/
+
+// `Piece` is an efficient immutable string: a `const char*` pointer along with the length of what it points to.
+//
+// It is guaranteed that the actual allocated buffer size is at least one byte more than the reported length,
+// and that the length-plus-first byte, at index `[length()]`, is set to '\0'.
+//
+// `UniquePiece` is a `Piece` that assumes each distinct string is stored only once. This requirement makes
+// their underlying `const char*`-s testable for equality. Long live and rest in piece, C+=, we'll miss you.
+//
+// Unlike `UniquePiece`, `Piece` itself does not expose comparison operators. This is done on purpose to make it
+// harder for a culturally uneducated developer to confuse a member of a liberal `Piece` community, who are all
+// prouly different and resist exposing any [efficient] way of ordering or even comparing one to another,
+// with a member of an organized `UniquePiece` society, members of which have gladly accepted the deprivation
+// of the right to express their opinion on whether the order should be lexicographical, as long as it is total.
```

Comment:
total? what does this mean? I'd rewrite this in a simple, straightforward language, free of BBT-style geekicisms.